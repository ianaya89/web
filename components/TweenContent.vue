<template>
  <div class="tween" :class="isVisible">
    <slot />
  </div>
</template>

<script>
import { setTimeout } from 'timers'
import { isScrolledIntoView } from '@/services/helpers'

export default {
  name: 'TweenContent',

  props: {
    localQuerySelector: {
      type: String,
      required: true
    },

    startValue: {
      type: Number,
      default: 0
    },

    duration: {
      type: Number,
      default: 1000
    },

    initVisible: {
      type: Boolean,
      default: false
    }
  },

  data: () => ({
    animated: false
  }),

  computed: {
    isVisible() {
      return this.animated || this.initVisible
        ? 'tween-visible'
        : 'tween-invisible'
    }
  },

  mounted() {
    if (!this.initVisible) {
      window.addEventListener('scroll', () => {
        if (!this.animated && isScrolledIntoView(this.$el)) {
          this.animateContent()
        }
      })
    } else if (!this.animated) {
      this.animateContent()
    }
  },

  methods: {
    animateContent() {
      const minTimer = 100
      const contentElement = this.$el.querySelector(this.localQuerySelector)
      const textStarting = contentElement.innerHTML
      const endValue = parseInt(textStarting.replace(/\D/g, ''))
      const range = endValue - this.startValue
      const duration = this.duration

      let stepTime = Math.abs(Math.floor(duration / range))
      stepTime = Math.max(stepTime, minTimer)

      const startTime = new Date().getTime()
      const endTime = startTime + duration
      let timer = null

      timer = setInterval(() => {
        const now = new Date().getTime()
        const remaining = Math.max((endTime - now) / duration, 0)
        const value = Math.round(endValue - remaining * range)
        contentElement.innerHTML = textStarting.replace(/([0-9]+)/g, value)
        if (value === endValue) {
          clearInterval(timer)
        }
      }, stepTime)

      setTimeout(() => {
        this.animated = true
      }, stepTime)
    }
  }
}
</script>

<style lang="postcss" scoped>
.tween {
  transition: opacity 1000ms linear;
  will-change: opacity;

  &.tween-invisible {
    opacity: 0;
  }
  &.tween-visible {
    opacity: 100;
  }
}
</style>
