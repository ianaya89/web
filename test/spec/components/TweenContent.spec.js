import { mount } from '@vue/test-utils'
import TweenContent from '@/components/TweenContent'

describe('TweenContent', () => {
  it('renders tween', () => {
    const wrapper = mount(TweenContent, {
      propsData: {
        localQuerySelector: 'span',
        duration: 500
      },
      slots: {
        default: '<span>1230 ks</span>'
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
