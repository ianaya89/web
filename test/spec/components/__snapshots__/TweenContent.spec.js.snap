// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TweenContent renders tween 1`] = `
<div
  class="tween tween-invisible"
>
  <span>
    1230 ks
  </span>
</div>
`;
